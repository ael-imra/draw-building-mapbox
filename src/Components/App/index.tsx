import { Grid } from '@mui/material';
import { useState } from 'react';
import { useMap } from '../../hooks';
import SettingsModal from '../SettingsModal';
import { SettingsActionKind, SettingsState } from '../SettingsModal/types';
import StatisticsModal from '../StatisticsModal';

const App = () => {
  const [geoJson, setGeoJson] = useState<any | null>(null);
  const { container, addFloors, resizeFloors, addConverage, statistics } = useMap(geoJson);

  const onSliderChange = (settings: SettingsState, type: SettingsActionKind) => {
    if (type === SettingsActionKind.STATE_FLOOR_NUMBER) addFloors(settings);
    if (type === SettingsActionKind.STATE_FLOOR_HEIGHT) resizeFloors(settings);
    if (type === SettingsActionKind.STATE_COVERAGE) addConverage(settings);
  };

  return (
    <>
      <Grid sx={{ width: '100%', height: '100%' }} ref={container} />
      <SettingsModal onDataChange={setGeoJson} onSliderChange={onSliderChange} />
      <StatisticsModal {...(statistics as any)} />
    </>
  );
};

export default App;
