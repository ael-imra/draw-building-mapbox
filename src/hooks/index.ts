import mapbox, { FillExtrusionLayer, GeoJSONSource, LineLayer, Map } from 'mapbox-gl';
import * as turf from '@turf/turf';
import { useEffect, useRef, useState } from 'react';
import { SettingsState } from '../Components/SettingsModal/types';
import { INITIAL_STATE } from '../Components/SettingsModal';

mapbox.accessToken = process.env.REACT_APP_MAP_KEY as string;

export const useMap = (geoJson: any) => {
  const map = useRef<Map | null>(null);
  const container = useRef<any | null>(null);

  const [statistics, setStatistics] = useState({});

  useEffect(() => {
    if (!map.current) {
      map.current = new mapbox.Map({
        container: container.current,
        style: 'mapbox://styles/brave123/cldlnbhah004c01t9f6j38t4j',
      });
    }
  }, []);

  useEffect(() => {
    if (geoJson && map.current) {
      clearSource();
      map.current.addSource('source_1', {
        type: 'geojson',
        data: geoJson,
      });
      const bbox = turf.bbox(geoJson);
      map.current.flyTo({
        center: [bbox[2], bbox[3]],
        zoom: 18,
        pitch: 60,
      });
      addConverage(INITIAL_STATE);
      addFloors(INITIAL_STATE);
      resizeFloors(INITIAL_STATE);
    }
    // eslint-disable-next-line
  }, [geoJson]);

  const addLayer = (props: LineLayer | FillExtrusionLayer) => {
    if (map.current && geoJson) {
      if (map.current.getLayer(props.id)) {
        if (props.paint) {
          for (const [key, value] of Object.entries(props.paint)) {
            map.current.setPaintProperty(props.id, key, value);
          }
        }
      } else map.current.addLayer(props);
    }
  };

  const addFloors = (settings: SettingsState) => {
    const { floorNumber, floorHeight } = settings;
    if (map.current && geoJson) {
      for (let i = 0; i < floorNumber; i++) {
        addLayer({
          id: `layer_${i}`,
          type: 'fill-extrusion',
          paint: {
            'fill-extrusion-color': '#f9d272',
            'fill-extrusion-height': floorHeight * (i + 1),
            'fill-extrusion-base': i * floorHeight + 1,
          },
          source: 'source_1',
        });
      }
      getStatistics(settings);

      const buildingHeight = floorHeight * floorNumber;
      if (buildingHeight > 0) {
        const zoom = 17 - Math.log2(buildingHeight / 150);
        map.current?.setZoom(zoom);
      }
    }
  };

  const resizeFloors = (settings: SettingsState) => {
    const { floorNumber, floorHeight } = settings;
    if (map.current && geoJson) {
      for (let i = 0; i < floorNumber; i++) {
        map.current.setPaintProperty(`layer_${i}`, 'fill-extrusion-height', floorHeight * (i + 1));
        map.current.setPaintProperty(`layer_${i}`, 'fill-extrusion-base', i * floorHeight + 1);
      }
      getStatistics(settings);
      const buildingHeight = floorHeight * floorNumber;
      if (buildingHeight > 0) {
        const zoom = 17 - Math.log2(buildingHeight / 150);
        map.current?.setZoom(zoom);
      }
    }
  };

  const addConverage = (settings: SettingsState) => {
    if (map.current && geoJson) {
      const bbox = turf.transformScale(geoJson, settings.coverage / 100 + 1);
      const source = map.current.getSource('coverage') as GeoJSONSource;
      if (source) source.setData(bbox);
      else
        map.current?.addSource('coverage', {
          type: 'geojson',
          data: bbox,
        });
      addLayer({
        id: 'coverage',
        type: 'fill-extrusion',
        paint: {
          'fill-extrusion-color': '#dcf0e1',
        },
        source: 'coverage',
      });
      addLayer({
        id: 'coverage_border',
        type: 'line',
        paint: {
          'line-color': '#5d605c',
          'line-width': 3,
          'line-dasharray': [7, 3, 4],
        },
        source: 'coverage',
      });
      getStatistics(settings);
    }
  };

  const clearSource = () => {
    if (map.current) {
      const { layers } = map.current.getStyle();
      for (const layer of layers)
        if (layer.id.indexOf('layer_') > -1 || layer.id.indexOf('coverage') > -1) map.current.removeLayer(layer.id);
      if (map.current.getSource('source_1')) map.current.removeSource('source_1');
      if (map.current.getSource('coverage')) map.current.removeSource('coverage');
    }
  };

  const getStatistics = (settings: SettingsState) => {
    if (geoJson) {
      const { floorHeight, floorNumber } = settings;

      const buildingArea = turf.round(turf.area(geoJson));
      const bboxTransformed = turf.transformScale(geoJson, settings.coverage / 100 + 1);
      const landArea = turf.round(turf.area(bboxTransformed));
      const height = floorHeight * floorNumber + floorNumber;
      setStatistics({
        landArea,
        buildingArea,
        buildingFloorArea: floorNumber * buildingArea,
        volume: buildingArea * height,
        buildingHeight: height,
      });
    }
  };

  return {
    map: map.current,
    container,
    statistics,
    addFloors,
    resizeFloors,
    addConverage,
  };
};
