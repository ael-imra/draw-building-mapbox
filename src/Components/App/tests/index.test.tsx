import { render, cleanup } from '@testing-library/react';
import * as hooks from '../../../hooks';
import App from '../';
import { Map } from 'mapbox-gl';

afterEach(cleanup);
describe('App', () => {
  it('should render correctly', () => {
    jest.spyOn(hooks, 'useMap').mockImplementation(() => ({
      addConverage: jest.fn(),
      addFloors: jest.fn(),
      container: { current: jest.fn() },
      map: {} as Map,
      resizeFloors: jest.fn(),
      statistics: {},
    }));
    const { asFragment } = render(<App />);

    expect(asFragment()).toMatchSnapshot();
  });
});
