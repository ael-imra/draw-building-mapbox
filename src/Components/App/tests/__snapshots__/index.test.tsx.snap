// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root css-1hj1l9r-MuiGrid-root"
  />
  <div
    class="MuiGrid-root css-vj1n65-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-container css-1fsqadj-MuiGrid-root"
    >
      <div
        class="MuiFormControl-root MuiTextField-root css-1qf7f8o-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            data-testid="file"
            id="file"
            type="file"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-ihdtdm"
            >
              <span
                class="notranslate"
              >
                â€‹
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiGrid-root css-1onjoz5-MuiGrid-root"
      >
        Load GeoJson
      </div>
      <div
        class="MuiGrid-root MuiGrid-container css-fvth8m-MuiGrid-root"
      >
        <div
          class="MuiGrid-root css-1y7x7wc-MuiGrid-root"
        >
          lot converage %
        </div>
        <div
          class="MuiGrid-root MuiGrid-container css-1u2cepy-MuiGrid-root"
        >
          <span
            class="MuiSlider-root MuiSlider-marked MuiSlider-colorPrimary MuiSlider-sizeMedium css-1fdx4to-MuiSlider-root"
          >
            <span
              class="MuiSlider-rail css-14pt78w-MuiSlider-rail"
            />
            <span
              class="MuiSlider-track css-1gv0vcd-MuiSlider-track"
              style="left: 0%; width: 0%;"
            />
            <span
              class="MuiSlider-mark MuiSlider-markActive css-6cwnna-MuiSlider-mark"
              data-index="0"
              style="left: 0%;"
            />
            <span
              aria-hidden="true"
              class="MuiSlider-markLabel MuiSlider-markLabelActive css-1eoe787-MuiSlider-markLabel"
              data-index="0"
              style="left: 0%;"
            >
              0
            </span>
            <span
              class="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
              data-index="1"
              style="left: 100%;"
            />
            <span
              aria-hidden="true"
              class="MuiSlider-markLabel css-yafthl-MuiSlider-markLabel"
              data-index="1"
              style="left: 100%;"
            >
              100
            </span>
            <span
              class="MuiSlider-thumb MuiSlider-thumbSizeMedium MuiSlider-thumbColorPrimary css-eg0mwd-MuiSlider-thumb"
              data-focusvisible="false"
              data-index="0"
              style="left: 0%; pointer-events: none;"
            >
              <input
                aria-orientation="horizontal"
                aria-valuemax="100"
                aria-valuemin="0"
                aria-valuenow="0"
                data-index="0"
                max="100"
                min="0"
                step="1"
                style="border: 0px; height: 100%; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 100%; direction: ltr;"
                type="range"
                value="0"
              />
              <span
                aria-hidden="true"
                class="MuiSlider-valueLabel css-nnid7-MuiSlider-valueLabel"
              >
                <span
                  class="MuiSlider-valueLabelCircle"
                >
                  <span
                    class="MuiSlider-valueLabelLabel"
                  >
                    0
                  </span>
                </span>
              </span>
            </span>
          </span>
          <div
            class="MuiGrid-root MuiGrid-container css-fvskug-MuiGrid-root"
          >
            0
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-container css-fvth8m-MuiGrid-root"
      >
        <div
          class="MuiGrid-root css-1y7x7wc-MuiGrid-root"
        >
          floor number
        </div>
        <div
          class="MuiGrid-root MuiGrid-container css-1u2cepy-MuiGrid-root"
        >
          <span
            class="MuiSlider-root MuiSlider-marked MuiSlider-colorPrimary MuiSlider-sizeMedium css-1fdx4to-MuiSlider-root"
          >
            <span
              class="MuiSlider-rail css-14pt78w-MuiSlider-rail"
            />
            <span
              class="MuiSlider-track css-1gv0vcd-MuiSlider-track"
              style="left: 0%; width: 0%;"
            />
            <span
              class="MuiSlider-mark MuiSlider-markActive css-6cwnna-MuiSlider-mark"
              data-index="0"
              style="left: 0%;"
            />
            <span
              aria-hidden="true"
              class="MuiSlider-markLabel MuiSlider-markLabelActive css-1eoe787-MuiSlider-markLabel"
              data-index="0"
              style="left: 0%;"
            >
              0
            </span>
            <span
              class="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
              data-index="1"
              style="left: 100%;"
            />
            <span
              aria-hidden="true"
              class="MuiSlider-markLabel css-yafthl-MuiSlider-markLabel"
              data-index="1"
              style="left: 100%;"
            >
              100
            </span>
            <span
              class="MuiSlider-thumb MuiSlider-thumbSizeMedium MuiSlider-thumbColorPrimary css-eg0mwd-MuiSlider-thumb"
              data-focusvisible="false"
              data-index="0"
              style="left: 0%; pointer-events: none;"
            >
              <input
                aria-orientation="horizontal"
                aria-valuemax="100"
                aria-valuemin="0"
                aria-valuenow="0"
                data-index="0"
                max="100"
                min="0"
                step="1"
                style="border: 0px; height: 100%; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 100%; direction: ltr;"
                type="range"
                value="0"
              />
              <span
                aria-hidden="true"
                class="MuiSlider-valueLabel css-nnid7-MuiSlider-valueLabel"
              >
                <span
                  class="MuiSlider-valueLabelCircle"
                >
                  <span
                    class="MuiSlider-valueLabelLabel"
                  >
                    0
                  </span>
                </span>
              </span>
            </span>
          </span>
          <div
            class="MuiGrid-root MuiGrid-container css-fvskug-MuiGrid-root"
          >
            0
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-container css-fvth8m-MuiGrid-root"
      >
        <div
          class="MuiGrid-root css-1y7x7wc-MuiGrid-root"
        >
          floor height
        </div>
        <div
          class="MuiGrid-root MuiGrid-container css-1u2cepy-MuiGrid-root"
        >
          <span
            class="MuiSlider-root MuiSlider-marked MuiSlider-colorPrimary MuiSlider-sizeMedium css-1fdx4to-MuiSlider-root"
          >
            <span
              class="MuiSlider-rail css-14pt78w-MuiSlider-rail"
            />
            <span
              class="MuiSlider-track css-1gv0vcd-MuiSlider-track"
              style="left: 0%; width: 0%;"
            />
            <span
              class="MuiSlider-mark MuiSlider-markActive css-6cwnna-MuiSlider-mark"
              data-index="0"
              style="left: 0%;"
            />
            <span
              aria-hidden="true"
              class="MuiSlider-markLabel MuiSlider-markLabelActive css-1eoe787-MuiSlider-markLabel"
              data-index="0"
              style="left: 0%;"
            >
              0
            </span>
            <span
              class="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
              data-index="1"
              style="left: 100%;"
            />
            <span
              aria-hidden="true"
              class="MuiSlider-markLabel css-yafthl-MuiSlider-markLabel"
              data-index="1"
              style="left: 100%;"
            >
              100
            </span>
            <span
              class="MuiSlider-thumb MuiSlider-thumbSizeMedium MuiSlider-thumbColorPrimary css-eg0mwd-MuiSlider-thumb"
              data-focusvisible="false"
              data-index="0"
              style="left: 0%; pointer-events: none;"
            >
              <input
                aria-orientation="horizontal"
                aria-valuemax="100"
                aria-valuemin="0"
                aria-valuenow="0"
                data-index="0"
                max="100"
                min="0"
                step="1"
                style="border: 0px; height: 100%; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 100%; direction: ltr;"
                type="range"
                value="0"
              />
              <span
                aria-hidden="true"
                class="MuiSlider-valueLabel css-nnid7-MuiSlider-valueLabel"
              >
                <span
                  class="MuiSlider-valueLabelCircle"
                >
                  <span
                    class="MuiSlider-valueLabelLabel"
                  >
                    0
                  </span>
                </span>
              </span>
            </span>
          </span>
          <div
            class="MuiGrid-root MuiGrid-container css-fvskug-MuiGrid-root"
          >
            0
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="MuiGrid-root MuiGrid-container css-1mt2vs5-MuiGrid-root"
  >
    <div
      class="MuiGrid-root css-18cuka8-MuiGrid-root"
    >
      Statistics
    </div>
    <div
      class="MuiGrid-root MuiGrid-container css-39fura-MuiGrid-root"
      data-testid="statisticsItem"
    >
      Land Area 
      <span />
    </div>
    <div
      class="MuiGrid-root MuiGrid-container css-39fura-MuiGrid-root"
      data-testid="statisticsItem"
    >
      Building Area 
      <span />
    </div>
    <div
      class="MuiGrid-root MuiGrid-container css-39fura-MuiGrid-root"
      data-testid="statisticsItem"
    >
      Building Floor Area 
      <span />
    </div>
    <div
      class="MuiGrid-root MuiGrid-container css-39fura-MuiGrid-root"
      data-testid="statisticsItem"
    >
      Volume 
      <span />
    </div>
    <div
      class="MuiGrid-root MuiGrid-container css-39fura-MuiGrid-root"
      data-testid="statisticsItem"
    >
      Building Height 
      <span />
    </div>
  </div>
</DocumentFragment>
`;
